### ISSUE TITLE
The performance is VERY different to sklearn.manifold.TSNE

### ISSUE THREAD

**Issue by elgartmi** [jsilter/parametric_tsne#6]
First of all, great work! :)
It's much needed for a bunch of reasons. 

Second, in the readme, you should add "ptSNE.fit(...)", as it is written now, it doesn't do anything. 

Third, I tried to compare this implementation to the sklearn one based on mnist, and this one performs rather poorly for me... I tried going up in the epochs as far as 400 (VERY slow, even on my GPUs), but not much improvement...
Is this my error in usage? 

Here is the code:

import numpy as np
np.random.seed(71)

import matplotlib
from keras import backend as K
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation, Flatten
from keras.layers.convolutional import Convolution2D, MaxPooling2D
from keras.optimizers import SGD
from keras.callbacks import Callback
from keras.utils import np_utils
from keras.objectives import categorical_crossentropy
from keras.datasets import cifar10, mnist

import matplotlib.pyplot as plt

# # mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()
n, row, col = X_train.shape
channel = 1

X_train = X_train.reshape(-1, channel * row * col)
X_test = X_test.reshape(-1, channel * row * col)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

#just make this run faster
X_train = X_train[:5000,:]
X_test = X_test[:500,:]
y_train = y_train[:500]
y_test = y_test[:500]

print("X_train.shape:", X_train.shape)
print("X_test.shape:", X_test.shape)

high_dims = X_train.shape[1]
num_outputs = 2
perplexity = 20
ptSNE = Parametric_tSNE(high_dims, num_outputs, perplexity)
ptSNE.fit(X_train, verbose=1,epochs=400)
output_res = ptSNE.transform(X_train)
output_res2 = ptSNE.transform(X_test)

plt.scatter(output_res[:, 0], output_res[:, 1], marker='o', s=4, edgecolor='')
plt.savefig("tsne_train_result.png")

plt.scatter(output_res2[:, 0], output_res2[:, 1], marker='o', s=4, edgecolor='')
plt.savefig("tsne_train_result2.png")

from sklearn import manifold
cdata_tsne0 = manifold.TSNE(n_components=2, init='random', random_state=0, perplexity=20,
                     learning_rate=300, n_iter=400)
print("Doing tSNE")
cdata_tsne_out0 = cdata_tsne0.fit_transform(X_train)
plt.scatter(cdata_tsne_out0[:,0], cdata_tsne_out0[:,1], marker='o', s=4, edgecolor='')
plt.savefig("tsne_train_result_orig.png")

from IPython.display import Image
Image("tsne_train_result1.png")
Image("tsne_train_result2.png")
Image("tsne_train_result_orig.png")

**Comment by jsilter** [comment]
Thanks for noticing the README error, I fixed that. I also added a link to Laurens van der Maatens tSNE FAQ, you may find useful advice there as well.

I haven't run your example yet, but frankly I would be surprised to see any meaningful results on MNIST characters just being unspooled into 1D. Any translation or rotation of characters is going to have dramatic effects on the 1D data vector

That said, you might try:
* Varying the perplexity parameters (it can also be an array of multiple values)
* Running PCA on the input data before putting it into tSNE and taking the first 50-100 dimensions. This should speed up training and provide better results. The sklearn.manifold.TSNE can do this with `init='pca'`; for my implementation you'll need to transform the input data but it shouldn't be too hard with scikits PCA tool.

**Comment by elgartmi** [comment]
Hi. Sure, you are correct (I've read the paper, using tSNE constantly everywhere etc).

The point is for mnist  -  sklearn tSNE works well (for 2D) producing ~meaningful results (see http://colah.github.io/posts/2014-10-Visualizing-MNIST/) , while this KERAS implementation doesn't (just basically gives one big blob). I was wondering just about this difference in performance. The mnist here is also irrelevant, just a convenient dataset. 

I want to use this code on my own data, so I'm wondering whether I'm using it wrong, or is there some other issue.

Thanks again!

**Comment by jsilter** [comment]
Your usage is correct, that shouldn't be an issue. I've run your example and see what you mean. I don't know how to explain the poor performance (I tried reducing the input data dimensionality via PCA, it doesn't look much different). It may be that the default neural network doesn't work well for this data, that can be changed via "all_layers".
